### ISSUE TITLE
Enhancement: replace `fast-glob` with `tinyglobby`

### ISSUE THREAD

**Issue by benmccann** [typescript-eslint/typescript-eslint#10533]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-estree

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

`fast-glob` has 17 dependencies with 12 people having publish access to those libraries and takes 505kb after installation
https://npmgraph.js.org/?q=fast-glob
https://pkg-size.dev/fast-glob

`tinyglobby` has only 2 dependencies with 6 people having publish access to those libraries and takes 155kb after installation
https://npmgraph.js.org/?q=tinyglobby
https://pkg-size.dev/tinyglobby

I would be happy to send a PR for this

### Additional Info

This was discussed a bit in https://github.com/typescript-eslint/typescript-eslint/issues/9453, but I thought it would be a good time to revisit as `tinyglobby` usage has grown 30x since that time having been adopted by projects like `vite`, `nx`, and `nuxt`. `tinyglobby` is likely to dedupe in user projects now. E.g. this repo already pulls in `tinyglobby` multiple times (via `nx` and `cspell`). And other projects in the eslint ecosystem like `eslint-import-resolver-typescript` (downloaded 10m times per week) depend on `tinyglobby` already. While `fast-glob` is still downloaded more frequently, 80% of those downloads come from this project, so whatever library this project uses will end up being the most used.

**Comment by james-pre** [comment]
@benmccann

I'm not a maintainer, but I think replacing `fast-glob` would be great. Going off of the replacement of `is-number` in https://github.com/micromatch/to-regex-range (which hasn't been released after **5 months**), I think this could save a few *Terabytes* of weekly traffic on npm, which is a huge win.

Just a thought, it may be worth-while to use [`fs.globSync`](https://nodejs.org/docs/v22.0.0/api/fs.html#fsglobsyncpattern-options) rather than another package- this has some pretty big benefits considering it completely eliminates the dependency. It would require a change to the Node.js version needed though (`>= 22.0.0`).

**Comment by bradzacher** [comment]
> it may be worth-while to use fs.globSync rather than another package- this has some pretty big benefits considering it completely eliminates the dependency. It would require a change to the Node.js version needed though (>= 22.0.0).

We cannot do that until our minimum version is v22 - which won't happen until v20 is out of LTS

**Comment by james-pre** [comment]
Thanks for the quick reply.

It looks like Node v20 is out of active LTS (as of Oct. 2024) and will be in maintenance until April 2026.

It was trivial for me to make the change to using the built-in `globSync`, so I went ahead and made the change locally. We can always put this on ice: https://github.com/james-pre/typescript-eslint/commit/d55558968f44bc3a9d514c66863d7d2651098d55.

I'll look into switching to `tinyglobby` tomorrow if I have time.

**Comment by JoshuaKGoldberg** [comment]
I'm mildly 👍 on this. It's a fine idea. I see no downside now that the library is widely depended-upon.

**Comment by ronami** [comment]
I think this is a good idea, having less dependencies is always nice.

However, I think it may take some time for `tinyglobby` to be as stable as `fast-glob`, and using it now may potentially introduce bugs (as in [1](https://github.com/dsherret/ts-morph/pull/1558#issuecomment-2558237500), [2](https://github.com/streetsidesoftware/cspell/pull/6167#issuecomment-2415942929), [3](https://github.com/vitest-dev/vitest/pull/6688)).

From what I can see, `tinyglobby` is getting more popular, and those bugs are getting fixed, so I think it depends how stable it currently is.

**Comment by JoshuaKGoldberg** [comment]
Oof, yeah, good call @ronami. I think we should wait a few months for this to all settle down.

**Comment by james-pre** [comment]
It looks like the problems referenced by @ronami have either been resolved or don't affect this project, since *all* of the tests in the draft PR passed.

**Comment by JoshuaKGoldberg** [comment]
True, but the fact that there were problems in the first place shakes confidence in the project.

This isn't a pressing issue IMO. It's nice to reduce dependency sizes, but given how many projects are still using `fast-glob` -including those linked ones- I don't think we're the only blocker to a bunch of npm savings.

**Comment by benmccann** [comment]
I think waiting a few months is fair as `tinyglobby` is still a somewhat new library.

I will say that the `tinyglobby` author has fixed any issues very quickly as they've been reported and has added a test each time to prevent any regressions. E.g. in that link to the cspell repo, the issue was promptly fixed and then [they upgraded to the latest](https://github.com/streetsidesoftware/cspell/pull/6419) rather than rolling back and vitest is also in process of switching back to `tinyglobby` again. I'm also talking to the `tinyglobby` author about seeing if we can automatically run the test suites of all the major projects that depend on it to instill more confidence as well (this is something we've done for Svelte and Vite).

Btw, `globby` / `fast-glob` are not without issues. I've hit multiple issues with it when switching where `tinyglobby` is returning something different and it's very clearly a bug in `globby` / `fast-glob`, but people have just gotten used to working around it. New bugs tend to be more noticeable than old bugs though.

Anyway, I think it makes sense for a project as large as this one to be more conservative and wait for any issues to shake out with other consumers first. Let's check back in a few months from now and see if everything has been quiet with others adopting the library.

**Comment by mrmlnc** [comment]
> Btw, `globby` / `fast-glob` are not without issues. I've hit multiple issues with it when switching where `tinyglobby` is returning something different and it's very clearly a bug in `globby` / `fast-glob`, but people have just gotten used to working around it. New bugs tend to be more noticeable than old bugs though.

To me, this looks like too aggressive marketing. In many issues and PR's mentioning `fast-glob`, I see your comment. New glob library come along every year, that has only a subset of the functionality of the old one and claims to be better. When these libraries start to be used, people start asking to add the same options that are available in `fast-glob`. In your case, it will be brace expansion, stream API, error handling, smart complex pattern processing when reading deeply, …. Then the maintainers face a choice – to write a solution themselves (without dependencies) or to use already existing and tested solutions (like `braces`, `merge2`, …).

I do not share the opinion that the number of dependencies (under reasonable assumptions) is a selling point. Stability, functionality and maintainability – yes. For me. Some rational balance is needed here. People may think differently.

> …it's very clearly a bug in `globby` / `fast-glob`

Please report any issues you find with `fast-glob`. Yes, I don't fix them as quickly as people would like, but the critical problems have long been resolved.

No negativity. Just my thoughts. Having competitors is a positive thing. That's how I started this project [fast-glob] in 2017. I was not happy with the speed, complexity and maintainability of the existing solutions.

**Comment by benmccann** [comment]
> In many issues and PR's mentioning fast-glob, I see your comment.

I think in general that many projects have more dependencies than they need because they haven't paid much attention to it or prioritized it. In my own projects, I've often been able to reduce the number of dependencies by 80-90% just by spending some time on it ([example](https://x.com/BenjaminMcCann/status/1839349949605236753), [example](https://x.com/BenjaminMcCann/status/1810698991820321028)). I'm not affiliated with `tinyglobby`, but I have sent PRs to many projects reducing their dependency count and switching them to `tinyglobby` as I think there's benefit to it. I think `fast-glob` is great and have nothing negative to say about it. I think most projects don't need everything it offers though and can get by with less.

> When these libraries start to be used, people start asking to add the same options that are available in fast-glob. In your case, it will be brace expansion, stream API, error handling, smart complex pattern processing when reading deeply, …. Then the maintainers face a choice – to write a solution themselves (without dependencies) or to use already existing and tested solutions (like braces, merge2, …).

Yes, I agree that `fast-glob` is far more feature filled than `tinyglobby`. Most projects don't need most of the features that `fast-glob` has. `tinyglobby` has committed to staying small and not adding many of the features in `fast-glob`. Some may be added to `tinyglobby`, but most won't.

> Note that install size simply isn't a concern for us - we're a local, dev-only tool.
> Is it safer and more secure?

Less code and fewer dependencies means less attack surface.

Just a few months ago [there was a malware](https://therecord.media/malware-found-in-npm-package-with-millions-of-weekly-downloads) that affected a package that eslint developers install. Reducing the number of dependencies by a good percentage reduces your risk of facing these issues by a fair amount.

Also, `braces` which gets pulled in by `fast-glob` had a [recent security issue](https://security.snyk.io/vuln/SNYK-JS-BRACES-6838727). Whether or not this posed real security risk to users, they're notified and then have to spend the time to figure out whether there's some risk they face, upgrade their project to a newer version, etc. If you drop a large percentage of your dependencies, you also reduce the time you spend dealing with these sorts of issues.

>  IF AND ONLY IF no other package in the repo's closure depends on fast-glob (which is unlikely - eg eslint-import-resolver-typescript and @next/eslint-plugin-next both depend on it and are widely used).

`eslint-import-resolver-typescript`, the more widely used of these packages, uses `tinyglobby` ([source](https://github.com/import-js/eslint-import-resolver-typescript/blob/c7e4ba1cc999c1e55a1558e8dfe93be2375c138b/package.json#L85))

**Comment by bradzacher** [comment]
> Just a few months ago there was a malware that affected a package that eslint developers install.

The article you linked is from 2021 - not "just a few months ago".
And it's about `UAParser.js` which most eslint users wouldn't install AFAIK (what popular eslint plugin is doing UA parsing??).

> Also, braces which gets pulled in by fast-glob  hada recent security issue.

Reading that issue:

> An attacker can cause the application to allocate excessive memory and potentially crash by sending imbalanced braces as input

Which is a non-issue for eslint usecases. Because the only user inputs are from a user's config. So worst case a user could craft a glob that crashes their eslint run...
It's the same as any REDOS "vuln" - a complete non-issue for eslint usecases.

> If you drop a large percentage of your dependencies, you also reduce the time you spend dealing with these sorts of issues. 

Does it? What makes `tinyglobby` immune to such issues? Does it do advanced fuzz testing or something to look for and catch such vulnerabilities before a release? I would say that it's no less vulnerable than any other package. One could argue it's more vulnerable because the maintainers are responsible for a wider breadth of code due to not using focused dependencies like `braces` and so they are more likely to accidentally introduce vulns.

> eslint-import-resolver-typescript, the more widely used of these packages, uses tinyglobby 

That version is yet to be published.
For the foreseeable future people are likely going to be on a version that uses `fast-glob`.

As an example - we currently have over 20 paths that `fast-glob` is transitively depended on in this repo. So if we remove our direct dep on it in `/typescript-estree` then our total install size **will not change**.

**Comment by benmccann** [comment]
> The article you linked is from 2021 - not "just a few months ago".

Ah, my mistake on the date. Google shows the date as 2024 in the search results and the article says "The incident was detected on Friday, October 22" without mentioning the year, so I didn't notice it was a few years ago. As for whether eslint repo installs it, it appears that it doesn't currently, but did at that time. I happened to have a copy of eslint on my disk, which is what I checked when I saw the article, and it turns out that [my fork](https://github.com/benmccann/eslint) is a few years old. `ua-parser-js` was installed in the eslint repo via karma at that time. My point was that even if a package is installed as a dev dependency, the number of dependencies it has still matters. And some eslint developers and contributors themselves very possibly could have been hit by such a malicious package given that eslint was installing this one. Note that I was referring to developers and contributors of eslint and not end users of eslint. My point isn't that eslint users have been hit by such an issue, but that these issues are common (rspack was hit just 2 weeks ago) and even may have affected eslint contributors.

![Screenshot from 2025-01-05 15-05-02](https://github.com/user-attachments/assets/f2ea0d28-53d5-45d6-8a59-561c5b2a65a1)

> It's the same as any REDOS "vuln" - a complete non-issue for eslint usecases

My point wasn't that it posed a security risk, but that users need to take the time to evaluate whether it does and handle it accordingly. And even if they do research it and come to the conclusion that there's no risk, many are still going to do an upgrade to make dependabot or a vulerability scanner quiet down.

> That version is yet to be published.

Sure, but if this project were to adopt `tinyglobby` it will have been published and users are likely to have updated it before they upgrade this library or to do so at the same time.

**Comment by james-pre** [comment]
I think its important to point out that `fast-glob`'s dependency tree includes `is-number`:

![fast-glob](https://github.com/user-attachments/assets/6fe268d3-4a3a-4787-bf40-241237de19eb)

(`fast-glob` -> `micromatch` -> `braces` -> `fill-range` -> `to-regex-range` -> `is-number`).

`is-number` was removed from `to-regex-range` back in **July** to save 440GB of weekly NPM traffic ([`to-regex-range`#17](https://github.com/micromatch/to-regex-range/pull/17)), but Jon Schlinkert (the maintainer with NPM publish permissions) is refusing to release this because of his personal opinions ([`to-regex-range`#19](https://github.com/micromatch/to-regex-range/issue/19)). This means `to-regex-range` will no longer receive updates, so any security issues would go unpatched.

From some quick research, it looks like Jon is actively trying to prevent projects from moving away from his packages. Additionally, Jon is a maintainer of over 1,500 packages ([npm](https://www.npmjs.com/~jonschlinkert?activeTab=packages)), most of which haven't been updated in over a decade. Even if he wanted to maintain these packages, I seriously doubt any programmer would have the time to. 8/17 (47%) of `fast-glob`'s dependencies are maintained by Jon.

I see numerous problems similar to the ones leading up to the [xzutils incident](https://en.wikipedia.org/wiki/XZ_Utils_backdoor). A package (`to-regex-range`) is very widely depended on, is not being released on time (or at all), and has frustration around the maintainer. While there isn't any malicious code *yet*, there is the real possibility of a supply chain attack.

#### But other popular projects, like *X* still use `fast-glob`

This can go in circles... project A doesn't switch to `tinyglobby` because project B is still using `fast-glob` and project B doesn't switch to `tinyglobby` because project A is still using `fast-glob`, and so on. This argument creates a deadlock, preventing meaningful changes to the dependency tree. Realistically, both `fast-glob` and `tinyglobby` are popular enough that it doesn't really matter. 

#### But `tinyglobby` isn't as stable as `fast-glob`, it may have bugs

In my draft PR, all of the checks passed (https://github.com/typescript-eslint/typescript-eslint/pull/10544/checks). Sure, some extreme edge case may not work, but `fast-glob` will also have extreme edge cases that don't work. While `tinyglobby` may have the occasional bug, it will get fixed quickly. While `fast-glob` does okay with releases, its dependencies are another story- 9/17 (53%) haven't been updated in **years**: [`reusify`](https://www.npmjs.com/package/reusify), [`queue-microtask`](https://www.npmjs.com/package/queue-microtask), [`run-parallel`](https://www.npmjs.com/package/run-parallel), [`is-extglob`](https://www.npmjs.com/package/is-extglob), [`is-glob`](https://www.npmjs.com/package/is-glob), [`glob-parent`](https://www.npmjs.com/package/glob-parent), [`merge2`](https://www.npmjs.com/package/merge2), [`to-regex-range`](https://www.npmjs.com/package/to-regex-range), [`is-number`](https://www.npmjs.com/package/is-number).

#### `fs.glob` and `fs.globSync`

`fs.glob` and `fs.globSync` were introduced in Node v22, which is the current "active" LTS. Even though Node v20 will be maintenance-only until it is deprecated next year, it is worth considering whether we can just use the native implementation instead of a dependency, which would have the best impact on supply chain security.

#### Wrapping up

As @mrmlnc is `fast-glob`'s maintainer, I feel like they are biased towards keeping `fast-glob`. I don't maintain any of the current or proposed dependencies, so I don't have anything to gain or lose. My only motivation is improving the project and its supply chain.

If all that wasn't enough, `tinyglobby` comes with some notable performance improvements. For example, it uses `fdir`, which "can easily crawl a directory containing 1 million files in < 1 second."

Overall, there are more risks associated with `fast-glob` than `tinyglobby`, especially in the long term.

It would be incredibly short-sighted to stay with `fast-glob` given the clear issues with `fast-globs` dependencies, the performance and package size improvements provided by `tinyglobby`, and most importantly the massive supply chain security impact of switching.



Please let me know if there are any *other* things concerning `fast-glob` or `tinyglobby` that I didn't address.

**Comment by SuperchupuDev** [comment]
> As an example - we currently have over 20 paths that `fast-glob` is transitively depended on in this repo. So if we remove our direct dep on it in `/typescript-estree` then our total install size **will not change**.

@bradzacher just FYI, I suppose that all those 20 paths come from dev dependencies, because according to `npmgraph`, `typescript-eslint` only uses `fast-glob` in one place in production deps (the ones users will install) which is in `@typescript-eslint/typescript-estree` and no other subdependency. https://npmgraph.js.org/?q=typescript-eslint#select=exact%3Afast-glob%403.3.3

**Comment by Josh-Cena** [comment]
I'm +1 to keeping `fast-glob` just because it's already in production and we have verified that "it works". I don't think the benefits offered here, while definitely desirable, outweigh the overhead of a migration and dealing with possible incompatibilities with existing usage. After we require Node 22+ we can just go to `FS.glob` since that seems like the same level of potential breakage as migrating to another library.

**Comment by bradzacher** [comment]
> but I suppose that all those 20 paths come from dev dependencies, because according to npmgraph, typescript-eslint only uses fast-glob in one place in production deps (the ones users will install) which is in @typescript-eslint/typescript-estree and no other subdependency. So install size would go down for users

My point wasn't that our transitive deps come from dev deps -- my point was that our repo functions pretty well as an example of a pretty idiomatic TypeScript repo -- it has webpack plugins to bundle react code, it uses knip, etc.

I.e. our repo has some 20+ transitive paths through which `fast-glob` is currently depended on - so it's likely that the average TS repo also has **_at least_** 5 or more similar transitive dependency paths. So removing it from our production dep closure would provide zero improvement for the average TS repo because they would still be installing `fast-glob` for the remaining transitive requirements.

As an example - I just looked at my company's lockfile and I could quickly see over 30 transitive dependency pathways to `fast-glob`. Funnily enough too - they have ZERO references to `tinyglobby` - which means that if we make this change we don't reduce the install size or the total dependency count for the repo -- we will INCREASE both numbers -- we'll regress the very metric we're looking to improve!

**Comment by bradzacher** [comment]
As @Josh-Cena said - we have a working solution right now.
If we have a truly better library because it is meaningfully faster for users - then I'm all for it.

But if the metric we're working to improve is purely:
- "dependency count"
- "disk install size"
- security by reducing dependency count

But these are a false metrics to optimise for because there are so many usages of `globby` and `fast-glob` that removing just our usage won't actually improve things for anyone and could actually regress these very metrics for some users.

What is the improvement that end users will actually see?

**Comment by benmccann** [comment]
Fair enough. We can create some benchmarks to show what speed differences there are. I'd like to wait for the next release of `tinyglobby` before doing so as the upcoming release is significantly faster.

**Comment by 43081j** [comment]
pwoh this thread got complicated, fast.

here's a few things you should note...

@bradzacher the following:

- this whole point of "removing package X is nothing in the grand scheme of things"
  - if we all thought like this, progress would never be made. thankfully, there's a large amount of people who do not think like this and have been working very hard to _incrementally_ clean up dependency trees. indeed, that means sometimes a dependency is still somewhere deep in your tree even if we removed it as our direct dependency - its a stepping stone. at some point, we'll remove it from the others too but you're fully aware its never going to be the case that someone turns up with PRs across the ecosystem all at once to do this. **every little saving helps, because the sum of them is no longer small.**
- is it faster?
  - yes. tinyglobby uses fdir under the hood which is probably the fastest FS traversal library right now (all benchmarks i've seen agree with this, but happy to be proven wrong and can raise that with the maintainer so he can make it the fastest again)
- `You realistically install our dependencies once and never again as we very rarely actively bump our required minimum ranges`
  - not quite given a large amount of typescript's consumers are CI builds which don't cache installs. yes, i clone a repo and probably keep that installed copy a while, but im nothing compared to the millions of CI builds that don't do that.

TSESLint being a 'dev tool' doesn't give us an excuse to slack when it comes to performance. we should be aiming to make it faster, lighter and more modern. that won't happen all at once, it will happen over time.

@Josh-Cena the following:

you're correct in that we should certainly switch TSESLint to `fs.glob` when possible. if there isn't a tracking issue for this already, it'd be worth opening one so once we meet the minimum engine requirement, we can make the switch.

**Comment by JounQin** [comment]
`eslint-plugin-import-x`'s maintainer here, we're using `@typescript-eslint/utils` on production, but as this issue, it includes some features we're actually not using at all, so I'm planning to change `@typescript-eslint/utils` as dev dependency only.

**Comment by bradzacher** [comment]
@JounQin it's also worth noting that so many of your plugin's users are likely TS users - so they'll be installing `@typescript-eslint/utils` via our plugin's dependency on it anyways. (I.e. By copy pasting some of our code you've _increased_ the total install size for your users rather than decrease!)

Regardless - https://github.com/typescript-eslint/typescript-eslint/issues/10383 would probably be of interest to you.

**Comment by JounQin** [comment]
#10383 would be preferred indeed, I just want to move forward on our side at least first.

**Comment by JoshuaKGoldberg** [comment]
Coming over from https://github.com/typescript-eslint/typescript-eslint/issues/11212#issuecomment-3010281265:

> We're going to take a pause on e18e-style suggestions to move to functionally equivalent new dependencies for the sake of performance/size improvements. We've spent a larger than expected amount of our team budget triaging these issues. Most of them have ended up being low-user-impact and/or blocked on necessary work in the new dependencies. We need to context switch back to work that will more directly & more impactfully improve performance for users.
>
> To anybody who wishes to propose a new dependency to us, please:
> 1. Wait until at least November 1st, 2025
> 2. Provide the information asked for in  https://github.com/typescript-eslint/typescript-eslint/issues/11303#issuecomment-2998011166 and https://e18e.dev/guide/advocacy.html

Speaking to the specific suggestion of replacing `fast-glob` with `tinyglobby`: that's included. Looking at the benefits called out in this issue:

* Size: I personally agree that this would be an improvement. Especially given the ecosystem advances noted in #11340 that many projects have moved from `fast-glob` to `tinyglobby`! But, as Brad noted, it's <1% of our total install size in isolation - and there are many other libraries that still do duplicate the dependency.
* Speed: typescript-eslint does not run very many globs. The [new tinyglobby benchmarks](https://github.com/SuperchupuDev/tinyglobby/pull/122) show that all of the projects are quite fast, and generally take <<<1ms to do what users need. There is almost (but not quite) zero motivation on speed alone to switch.

So, while the net benefit is non-zero, for typescript-eslint, right now, it's outweighed by the drawbacks:

* Confidence: every change risks bugs; changes that touch file system checks and wildcard expansions even more so
* Maintenance time: other improvements such as #10383 and #10940 are likely to yield much better user-facing improvements per time spent
* Stability: `tinyglobby` is still new (or, perhaps now, not-old) and actively developing: we don't want to be an early (or, perhaps now, mid-stage) adopter

Again, we're not saying this change isn't valuable. We just can't justify spending our limited maintenance time on it right now. Let's re-evaluate and see where things are in November.

**Comment by benmccann** [comment]
Makes sense. I just wanted to share one more point for consideration, which is that while `fast-glob` is a small percentage of total install size, it could be as much as 30% of supply chain risk since 12 individuals have access to publish it or one of its dependencies (out of 40 total individuals for `typescript-eslint` and its dependencies).

**Comment by Kombuchelada** [comment]
fast-glob has gotten some negative attention lately. Some projects may be forced to drop typescript-eslint as a result. https://www.theregister.com/2025/08/27/popular_nodejs_utility_used_by/

**Comment by bradzacher** [comment]
@Kombuchelada In what world would anyone see the fact that the maintainer lives in russia as a vulnerability? How is that any more vulnerable than any other package on NPM? That line of thinking is absolutely ludicrous and that article is spreading completely baseless FUD.

If they ever introduced a backdoor into the tool it would immediately raise red flags, trigger CVEs and cause NPM to intervene to remove the vulnerable version(s) and likely remove the author's access. I.e. it would be handled just like any other breach that occurs (see the recent NX breach, for example, https://github.com/nrwl/nx/security/advisories/GHSA-cxm3-wv7p-598c)

The one and only source for that article is "Hunted Labs" -- a company who I'm sure is benefiting from the exposure and who is also infamously anti-open-source. They have an agenda they are fuelling with this article.

**Comment by Tyler8oliver** [comment]
Is this being prioritized? I'll need to drop the usage of this package and look for an alternative due to the negative press [here](https://www.theregister.com/2025/08/27/popular_nodejs_utility_used_by/)

**Comment by 43081j** [comment]
The press is fairly farfetched and more of an attempt at promoting a product than being actual news. Let's not base the migration request in this repo on it. 

Elsewhere, feel free to do whatever. But here we should focus on the other aspects I think

**Comment by Tyler8oliver** [comment]
@43081j I understand your reasoning and in 90+% of software products this isn't a concern but there are some industries where this press report is a major liability.

**Comment by Josh-Cena** [comment]
Let it be the team's opinion, as Brad already said above, that we do not care about that report or any of its repercussions; we are open to dropping `fast-glob`, but we will be migrating to a native solution instead of any other library, and the priority is entirely uninfluenced by the said article. This change also involves a breaking change, so it is unlikely to happen soon. Our code is MIT-licensed so we do not carry the liability of supporting high-sensitivity industries. If it matters to you, our code is also free to fork for yourself.

**Comment by bradzacher** [comment]
Echoing that - we will **_not_** be moving to replace the dependency due to the article, which we see as baseless, anti-open source, unnecessary fear mongering, and (frankly) rooted in racism.

We trust that NPM, GitHub and security researchers would catch and prevent backdoors added to `fast-glob` - just like they would do if **_any_** package had a back-door added - a prime example being the recent NX vulnerability https://github.com/nrwl/nx/security/advisories/GHSA-cxm3-wv7p-598c.

If you are worried about the vulnerability you should ensure you're using a lock file to pin your dependencies. If you want additional assurance then you can use features like [pnpm's overrides](https://pnpm.io/9.x/package_json#pnpmoverrides) or [npm's overrides](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#overrides) to guarantee specific versions of packages.

If you need to drop us (and thus all of eslint) as a dependency because of this article - we're sorry to see you go - but we will not be dropping the dependency any time soon.

**Comment by Kombuchelada** [comment]
All of the comments about the silliness of the article, and the reality of open source software are sound, valid, and I ultimately agree with them. Unfortunately for government projects, the decision to drop any dependency on `fast-glob` may not be up to individual projects or developers, and instead may come as an order from a less technical person higher up in the chain. 

The reason I made the comment was that I hoped in light of the recent bad press (and the community's general movement towards solutions like `tinyglob`) there was a chance of putting a little more priority on this issue. I totally understand that it doesn't matter for 90% or more of users. 

I would hate if we had to drop linting from our projects, but I understand if the maintainers can't prioritize this change.

**Comment by benmccann** [comment]
It's now November 1st, so I wanted to revisit this discussion. I hope you all had a Happy Halloween!

## High adoption enables deduplication

Of course `globby` and `fast-glob` are widely used so you may question whether switching a library to use `tinyglobby` just results in more dependencies rather than any savings once dependency deduplication is factored in.

A majority of the top 100 consumers of `fast-glob` and `globby` have now fully switched to `tinyglobby`. Of those that haven't switched, 8 are officially deprecated or legacy and a handful of others are unmaintained with no commits or releases for years, so a very large majority of actively maintained libraries have switched.
	
For example, a whole raft of popular build tools now rely solely on tinyglobby for glob functionality:

- Vite
- SWC
- copy-webpack-plugin
- tsup
- tsdown
- unbuild
- nx
- lerna
- and many more...

None of these tools have a deep dependency on other glob libraries - all of them use tinyglobby all the way down. 

Similarly, many popular frameworks have made the same move:

- React Router
- Preact
- Angular
- SvelteKit
- Astro
- Starlight
- Eleventy

Newly created apps using these frameworks no longer have transitive dependencies on any glob libraries other than `tinyglobby`

For others like Nuxt, SolidStart, and TanStack Start, all dependencies have fully switched to `tinyglobby` and those frameworks will use solely `tinyglobby` once they update their dependencies to the latest versions. Most remaining frameworks (i.e. Next) use both `tinyglobby` and `fast-glob`.

Many ecosystems now rely almost entirely on `tinyglobby` — you can now setup a new SvelteKit project with [every integration it offers](https://svelte.dev/docs/cli/sv-add#Official-add-ons) (`drizzle`, `eslint`, `lucia`, `mdsvex`, `paraglide`, `playwright`, `prettier`, `storybook`, `tailwindcss`, `vitest`, and all SvelteKit adapters) and your project will exclusively use `tinyglobby` - unless you use chose to use typescript and eslint in which case `typescript-eslint` pulls in `fast-glob`.

For projects that are new or have recently updated dependencies, `tinyglobby` is more likely to be used than `fast-glob` or `globby`. On both a relative and absolute basis, `tinyglobby` usage is growing faster than alternatives. Something important to keep in mind here is that all usage of `tinyglobby` is on a single release line where as it's split across major versions for alternatives. E.g. `globby` usage is split across ten versions (v5-14 all have more than 1m downloads/week). This means that the rate at which `tinyglobby` is being used on new projects compared to alternatives is even higher than raw download numbers might indicate.

## Supply chain risk

- `fast-glob` consists of [18 packages](https://npmgraph.js.org/?q=fast-glob) with an [install size of 513KB](https://pkg-size.dev/fast-glob)
- `tinyglobby` consists of [3 packages](https://npmgraph.js.org/?q=tinyglobby) with an [install size of 162KB](https://pkg-size.dev/tinyglobby)

Having fewer packages reduces the need to have as many people with access to merge and release new versions thus reducing supply chain vulnerabilities, of which we've seen a number of in the npm ecosystem. For example, just recently, users of `glob` were affected by a supply chain attack in which a dependency - `strip-ansi` - [was compromised](https://socket.dev/blog/npm-author-qix-compromised-in-major-supply-chain-attack). Users of `tinyglobby` would not have been affected by this particular attack and face far less risk of similar attacks in the future that other glob libraries do.

And for most new and actively updated projects, this isn't just 18 packages vs 3 packages, but 18 new packages vs 3 packages already in your dependency tree.

## Performance

This isn't meant to demonstrate that `tinyglobby` will overall improve performance for typescript-eslint users as globbing might be a small portion of the overall performance costs of running typescript-eslint, but it is generally the fastest available globbing library that I'm aware of. Here are benchmarks taken from the last commit on the `tinyglobby` CI:

```
Run node --run bench
packages/*/tsconfig.json (typescript-eslint)
┌─────────┬────────────────┬───────────────────┬───────────────────┬────────────────────────┬────────────────────────┬─────────┐
│ (index) │ Task name      │ Latency avg (ns)  │ Latency med (ns)  │ Throughput avg (ops/s) │ Throughput med (ops/s) │ Samples │
├─────────┼────────────────┼───────────────────┼───────────────────┼────────────────────────┼────────────────────────┼─────────┤
│ 0       │ 'tinyglobby'   │ '456457 ± 1.49%'  │ '427974 ± 16605'  │ '2260 ± 0.50%'         │ '2337 ± 93'            │ 2191    │
│ 1       │ 'fast-glob'    │ '579812 ± 2.54%'  │ '543256 ± 32441'  │ '1813 ± 0.60%'         │ '1841 ± 111'           │ 1725    │
│ 2       │ 'glob'         │ '612492 ± 3.46%'  │ '562362 ± 29505'  │ '1731 ± 0.61%'         │ '1778 ± 96'            │ 1633    │
│ 3       │ 'node:fs glob' │ '1170626 ± 1.09%' │ '1157149 ± 73148' │ '869 ± 0.76%'          │ '864 ± 55'             │ 855     │
└─────────┴────────────────┴───────────────────┴───────────────────┴────────────────────────┴────────────────────────┴─────────┘
**/* (typescript-eslint)
┌─────────┬────────────────┬─────────────────────┬───────────────────────┬────────────────────────┬────────────────────────┬─────────┐
│ (index) │ Task name      │ Latency avg (ns)    │ Latency med (ns)      │ Throughput avg (ops/s) │ Throughput med (ops/s) │ Samples │
├─────────┼────────────────┼─────────────────────┼───────────────────────┼────────────────────────┼────────────────────────┼─────────┤
│ 0       │ 'tinyglobby'   │ '24789520 ± 3.19%'  │ '23973377 ± 644311'   │ '41 ± 2.24%'           │ '42 ± 1'               │ 64      │
│ 1       │ 'fast-glob'    │ '50366452 ± 3.48%'  │ '50897896 ± 6895020'  │ '20 ± 3.33%'           │ '20 ± 3'               │ 64      │
│ 2       │ 'glob'         │ '60154522 ± 2.01%'  │ '60383243 ± 2965111'  │ '17 ± 2.06%'           │ '17 ± 1'               │ 64      │
│ 3       │ 'node:fs glob' │ '208721426 ± 1.68%' │ '207200996 ± 8277252' │ '5 ± 1.66%'            │ '5 ± 0'                │ 64      │
└─────────┴────────────────┴─────────────────────┴───────────────────────┴────────────────────────┴────────────────────────┴─────────┘
```

Hopefully this takes performance off the table as being a concern, so that you can at least be confident performance is unlikely to get worse.


## Stability

`tinyglobby` is a newer library, so you may ask whether switching to it introduces a higher risk of bugs. It is true that libraries take time to mature and in tinyglobby's earlier days some projects faced regressions. However, those issues have all now been fixed. With such widespread adoption by high profile widely-used projects in the ecosystem, any bugs in `tinyglobby` are found and reported quickly. Each issue has been patched with an included test and as a result `tinyglobby` now has over a hundred individual tests. When there has been a bug in `tinyglobby`, it's been relatively straightforward to identify whether it's in `tinyglobby` itself or one of its two dependencies. It can be much harder to track down an issue in a library with 17 dependencies or even to see all of the issues across all of the repos that might be present. In fact, `fast-glob` and its dependencies together have roughly twice as many open issues as `tinyglobby` and its dependencies combined. `tinyglobby` has been very actively maintained and has the support of the e18e community, so there has always been multiple contributors willing to look at issues.

`tinyglobby` now has 37m downloads/week coming from well known projects like vite, pnpm, and node-gyp showing that it has been widely adopted. With the majority of widely used actively maintained projects in the JS ecosystem having switched, `typescript-eslint` would not be an early or even mid-stage adopter.

**Comment by 43081j** [comment]
Since the only remaining reason to not move would be time constraints of the team, maybe it's worth opening the draft pr to get the ball rolling if nobody has yet. 

The more we can do now, the less the team needs to do

**Comment by james-pre** [comment]
I think we should just replace it with built-in globbiing functionality (i.e. `fs.globSync`)

**Comment by 43081j** [comment]
That would be the ideal but isn't available before node 22 iirc

We can get a win today by moving to tinyglobby, and later move to the built in functionality

It seems like a loss to give the users another year or so of a less performant library while we wait

**Comment by bradzacher** [comment]
> I think we should just replace it with built-in globbiing functionality (i.e. fs.globSync)

As James said -- the API is not available before v22.0 and was not made stable until v22.17.0 / v24.1.0.
ESLint v10 will drop support for <v20.19.0 -- so if we follow their lead we would be locked out of the API for a while. But Node v20 falls out of LTS April 2026 so we could align a major with that to allow us to jump to v22.17+. That pushes using the API out till _at least_ April 2026 (5 months).

**Comment by Josh-Cena** [comment]
Given that a dependency swap will be ~100LOC (from https://github.com/typescript-eslint/typescript-eslint/pull/10544), I'm +1 for changing even if it will just last for a few months (but who knows? we can get lazy and linger on with tinyglobby for quite a while)

**Comment by JoshuaKGoldberg** [comment]
Agreed, I'm also in favor of going with `tinyglobby` in this major and looking to move to native Node.js support in the next major >= April 2026.

Thanks e18e folks for doing such a good job of both pushing the library out in the ecosystem and justifying+validating it! ❤️

**Comment by benmccann** [comment]
Thank you guys for staying open to considering this. I've sent a PR for it: https://github.com/typescript-eslint/typescript-eslint/pull/11740. (the CI is red, but that's from a failing integration test on `main`. everything else is passing)

**Comment by benmccann** [comment]
Now that the CI is green on `main`, I've rebased the PR and it's green as well

**Comment by bradzacher** [comment]
Landed in https://github.com/typescript-eslint/typescript-eslint/pull/11740
